samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java

```diff
@@ -776,38 +810,67 @@ private List<String> getAttachmentResourceIds(NodeList list) {
 	
 	private <T> boolean migrateText(AssessmentService assessmentService, String toContext, T item, String itemHash,
 			boolean hasCaches,boolean hasDuplicates, boolean copyAttachments, String cacheCode, Map<String, String> textCache,
-			Set<Entry<String, String>> entrySet, Function<T, String> getter, BiConsumer<T, String> setter) {
+			Set<Entry<String, String>> entrySet, Map<String, String> transversalMap, MergeConfig mcx,
+			Function<T, String> getter, BiConsumer<T, String> setter) {
 
+		log.debug("migrateText: {} {}", itemHash, copyAttachments);
 		String cacheKey = itemHash + "-" + cacheCode;
 
 		if (hasCaches && textCache.containsKey(cacheKey)) {
-			// Item instruction has been cashed, lets get it form the cache
+			// Item instruction has been cached, lets get it from the cache
 			setter.accept(item, textCache.get(cacheKey));
 			return true;
 		} else {
 			// Item instruction has not been cached, lets try migrating
 			String itemText = StringUtils.trimToEmpty(getter.apply(item));
+
+			String migratedText = migrateTextString(assessmentService, toContext, itemHash, hasCaches, hasDuplicates,
+				copyAttachments, cacheCode, textCache, entrySet, transversalMap, mcx, itemText);
+
+			if (!StringUtils.equals(itemText, migratedText)) {
+				setter.accept(item, migratedText);
+				return true;
+			}
+		}
+		return false;
+	}
+
+	private String migrateTextString(AssessmentService assessmentService, String toContext, String itemHash,
```

Comment:
I don't fully understand this yet. It looks like you added some code to the part where the migration happens and then moved that into it's own method. But why does that also include the caching part? We would be double checking the caches?
If I understand the changes right, I think `migrateTextString` should just migrate without checking the cache and `migrateText` would call it after checking caches.

Reply:
Thanks - I will look into that.

Reply:
Actually I will move all caching into `migrateTextString()`.  The way I first did it avoided allocating one string that would be quickly deallocated.  But making the code simpler and clearer and not duplicating logic is more important.  Now `migrateText()` relies on `migrateTextString()` completely including all cache processing except for the check for differences, and using the setter and getter.  I wanted the instruction text to be cached as well.

Reply:
If you move caching to `migrateTextString` what purpose is left for `migrateText`? Why do we actually need two methods?
Either way, please ping me when you want me to look at it again.

Reply:
Good point.  I will go back to one migrate text function.